File: python/enolib/elements/field.py
from ..errors.validation import Validation
from . import attribute as attribute_module
from . import item as item_module
from . import section
from .missing import missing_attribute
from .value_element_base import ValueElementBase


class Field(ValueElementBase):
    def __init__(self, context, instruction, parent=None):
        super().__init__(context, instruction, parent)

        self._all_attributes_required = parent._all_elements_required if parent else False
    
    def __repr__(self):
        if 'attributes' in self._instruction:
            return f"<class Field key={self._instruction['key']} attributes={len(self._attributes())}>"
        elif 'items' in self._instruction:
            return f"<class Field key={self._instruction['key']} items={len(self._items())}>"
        elif 'continuations' in self._instruction or 'value' in self._instruction:
            return f"<class Field key={self._instruction['key']} value={self._print_value()}>"
            
        return f"<class Field key={self._instruction['key']}>"

    def _attribute(self, key, *, required=None):
        self._touched = True

        if not key:
            attributes = self._attributes()
        else:
            attributes_map = self._attributes(True)
            attributes = attributes_map[key] if key in attributes_map else []

        if len(attributes) == 0:
            if required or required is None and self._all_attributes_required:
                raise Validation.missing_element(self._context, key, self._instruction, 'missing_attribute')
            elif required is None:
                return missing_attribute.MissingAttribute(key, self)
            else:
                return None

        if len(attributes) > 1:
            raise Validation.unexpected_multiple_elements(
                self._context,
                key,
                [e._instruction for e in attributes],
                'expected_single_attribute'
            )

        return attributes[0]

    def _attributes(self, as_map=False):
        if not hasattr(self, '_instantiated_attributes'):
            self._instantiated_attributes_map = {}
            
            if 'attributes' in self._instruction:
                def instantiate_and_index(attribute):
                    instance = attribute_module.Attribute(self._context, attribute, self)

                    if attribute['key'] in self._instantiated_attributes_map:
                        self._instantiated_attributes_map[attribute['key']].append(instance)
                    else:
                        self._instantiated_attributes_map[attribute['key']] = [instance]

                    return instance

                self._instantiated_attributes = [instantiate_and_index(attribute) for attribute in self._instruction['attributes']]
            else:
                self._instantiated_attributes = []

        return self._instantiated_attributes_map if as_map else self._instantiated_attributes

    def _items(self):
        if not hasattr(self, '_instantiated_items'):
            if 'items' in self._instruction:
                self._instantiated_items = [item_module.Item(self._context, item, self) for item in self._instruction['items']]
            else:
                self._instantiated_items = []

        return self._instantiated_items

    def _missing_error(self, attribute):
        raise Validation.missing_element(self._context, attribute._key, self._instruction, 'missing_attribute')

    def _untouched(self):
        if not hasattr(self, '_touched'):
            return self._instruction

        if 'attributes' in self._instruction:
            untouched_attribute = next((attribute for attribute in self._attributes() if not hasattr(attribute, '_touched')), None)
            if untouched_attribute:
                return untouched_attribute._instruction
        elif 'items' in self._instruction:
            untouched_item = next((item for item in self._items() if not hasattr(item, '_touched')), None)
            if untouched_item:
                return untouched_item._instruction
            
        return False
    
    def all_attributes_required(self, required=True):
        self._all_attributes_required = required

    def assert_all_touched(self, message=None, *, only=None, skip=None):
        attributes_map = self._attributes(True)

        for key, attributes in attributes_map.items():
            if (skip and key in skip) or (only and key not in only):
                continue

            for attribute in attributes:
                if not hasattr(attribute, '_touched'):
                    if callable(message):
                        message = message(attribute)

                    raise Validation.unexpected_element(self._context, message, attribute['instruction'])

    def attribute(self, key=None):
        return self._attribute(key)

    def attributes(self, key=None):
        self._touched = True
        
        if 'continuations' in self._instruction or 'items' in self._instruction or 'value' in self._instruction:
            raise Validation.unexpected_field_content(self._context, key, self._instruction, 'expected_attributes')

        if not key:
            return self._attributes()
        else:
            attributes_map = self._attributes(True)
            return attributes_map[key] if key in attributes_map else []

    def items(self):
        self._touched = True
        
        if 'attributes' in self._instruction or 'continuations' in self._instruction or 'value' in self._instruction:
            raise Validation.unexpected_field_content(self._context, None, self._instruction, 'expected_items')

        return self._items()

    def length(self):
        self._touched = True

        return len(self._items())

    def optional_attribute(self, key=None):
        return self._attribute(key, required=False)

    def optional_string_values(self):
        self._touched = True

        return [item.optional_string_value() for item in self._items()]
        
    def optional_values(self, loader):
        self._touched = True

        return [item.optional_value(loader) for item in self._items()]

    def parent(self):
        return self._parent or section.Section(self._context, self._instruction['parent'])

    def required_attribute(self, key=None):
        return self._attribute(key, required=True)
    
    def required_string_values(self):
        self._touched = True

        if 'attributes' in self._instruction or 'continuations' in self._instruction or 'value' in self._instruction:
            raise Validation.unexpected_field_content(self._context, key, self._instruction, 'expected_items')

        return [item.required_string_value() for item in self._items()]
        
    def required_values(self, loader):
        self._touched = True

        if 'attributes' in self._instruction or 'continuations' in self._instruction or 'value' in self._instruction:
            raise Validation.unexpected_field_content(self._context, key, self._instruction, 'expected_items')

        return [item.required_value(loader) for item in self._items()]        
        
    def touch(self):
        self._touched = True

        if 'attributes' in self._instruction:
            for attribute in self._attributes():
                attribute._touched = True
        elif 'items' in self._instruction:
            for item in self._items():
                item._touched = True

File: ruby/lib/enolib/elements/field.rb
# frozen_string_literal: true

module Enolib
  class Field < ValueElementBase
    attr_reader :instruction, :touched
    
    def initialize(context, instruction, parent = nil)
      super(context, instruction, parent)

      @all_attributes_required = parent ? parent.all_elements_required? : false
    end
    
    def _missing_error(attribute)
      raise Errors::Validation.missing_element(@context, attribute.key, @instruction, 'missing_attribute')
    end
    
    def _untouched
      return @instruction unless instance_variable_defined?(:@touched)

      if @instruction.has_key?(:attributes)
        untouched_attribute = _attributes.find { |attribute| !attribute.instance_variable_defined?(:@touched) }
        return untouched_attribute.instruction if untouched_attribute
      elsif @instruction.has_key?(:items)
        untouched_item = _items.find { |item| !item.instance_variable_defined?(:@touched) }
        return untouched_item.instruction if untouched_item
      end
      
      false
    end
    
    def all_attributes_required(required = true)
      @all_attributes_required = required
    end
    
    def assert_all_touched(message = nil, except: nil, only: nil, &block)
      message = Proc.new(&block) if block_given?

      _attributes(map: true).each do |key, attributes|
        next if except && except.include?(key) || only && !only.include?(key)

        attributes.each do |attribute|
          next if attribute.touched  # TODO: Revisit instance var existance question in ruby

          if message.is_a?(Proc)
            message = message.call(attribute)
          end

          raise Errors::Validation.unexpected_element(@context, message, attribute[:instruction])
        end
      end
    end
    
    def attribute(key = nil)
      _attribute(key)
    end
    
    def attributes(key = nil)
      @touched = true

      if key
        attributes_map = _attributes(map: true)
        attributes_map.has_key?(key) ? attributes_map[key] : []
      else
        _attributes
      end
    end
    
    def items
      @touched = true

      _items
    end

    def length
      @touched = true

      _items.length
    end
    
    def optional_attribute(key)
      _attribute(key, required: false)
    end
    
    def optional_comment(loader = nil, &block)
      loader = Proc.new(&block) if block_given?

      if loader
        _comment(loader, required: false)
      else
        raise ArgumentError, 'A loader block or Proc must be provided'
      end
    end
    
    def optional_string_value
      _value(required: false)
    end
    
    def optional_string_values
      @touched = true

      _items.map(&:optional_string_value)
    end

    def optional_value(loader = nil, &block)
      loader = Proc.new(&block) if block_given?

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _value(loader, required: false)
    end
    
    def optional_values(loader = nil, &block)
      loader = Proc.new(&block) if block_given?

      @touched = true

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _items.map { |item| item.optional_value(loader) }
    end

    def parent
      @parent || Section.new(@context, @instruction[:parent])
    end
    
    def required_attribute(key = nil)
      _attribute(key, required: true)
    end

    def required_string_value
      _value(required: true)
    end
    
    def required_string_values
      @touched = true

      _items.map(&:required_string_value)
    end

    def required_value(loader = nil, &block)
      loader = Proc.new(&block) if block_given?

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _value(loader, required: true)
    end

    def required_values(loader = nil, &block)
      loader = Proc.new(&block) if block_given?

      @touched = true

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _items.map { |item| item.required_value(loader) }
    end

    def to_s
      if @instruction.has_key?(:attributes)
        "#<Enolib::Field key=#{@instruction[:key]} attributes=#{_attributes.length}>"
      elsif @instruction.has_key?(:items)
        "#<Enolib::Field key=#{@instruction[:key]} items=#{_items.length}>"
      elsif @instruction.has_key?(:continuations) || @instruction.has_key?(:value)
        "#<Enolib::Field key=#{@instruction[:key]} value=#{print_value}>"
      else
        "#<Enolib::Field key=#{@instruction[:key]}>"
      end
    end
    
    def touch
      @touched = true
      
      if @instruction.has_key?(:attributes)
        _attributes.each(&:touch)
      elsif @instruction.has_key?(:items)
        _items.each(&:touch)
      end
    end
    
    private
    
    def _attribute(key = nil, required: nil)
      @touched = true

      if key
        attributes_map = _attributes(map: true)
        attributes = attributes_map.has_key?(key) ? attributes_map[key] : []
      else
        attributes = _attributes
      end

      if attributes.empty?
        if required || required == nil && @all_attributes_required
          raise Errors::Validation.missing_element(@context, key, @instruction, 'missing_attribute')
        elsif required == nil
          return MissingAttribute.new(key, self)
        else
          return nil
        end
      end

      if attributes.length > 1
        raise Errors::Validation.unexpected_multiple_elements(
          @context,
          key,
          attributes.map(&:instruction),
          'expected_single_attribute'
        )
      end

      attributes[0]
    end
    
    def _attributes(map: false)
      unless instance_variable_defined?(:@instantiated_attributes)
        @instantiated_attributes_map = {}
        if @instruction.has_key?(:attributes)
          @instantiated_attributes = @instruction[:attributes].map do |attribute|
            instance = Attribute.new(@context, attribute, self)

            if @instantiated_attributes_map.has_key?(attribute[:key])
              @instantiated_attributes_map[attribute[:key]].push(instance)
            else
              @instantiated_attributes_map[attribute[:key]] = [instance]
            end

            instance
          end
        else
          @instantiated_attributes = []
        end
      end

      map ? @instantiated_attributes_map : @instantiated_attributes
    end
    
    def _items
      unless instance_variable_defined?(:@instantiated_items)
        @instantiated_items =
          if @instruction.has_key?(:items)
            @instruction[:items].map { |item| Item.new(@context, item, self) }
          else
            []
          end
      end

      @instantiated_items
    end
  end
end
